File: gesturePad/Plugins/VolumeSlider.m
//
//  	VolumeSlider.m
//  	Volume Slider Cordova Plugin
//
//  	Created by Tommy-Carlos Williams on 20/07/11.
//  	Copyright 2011 Tommy-Carlos Williams. All rights reserved.
//      MIT Licensed
//

#import "VolumeSlider.h"

@implementation VolumeSlider

@synthesize mpVolumeViewParentView, myVolumeView, callbackId;

#ifndef __IPHONE_3_0
@synthesize webView;
#endif


-(CDVPlugin*) initWithWebView:(UIWebView*)theWebView
{
    self = (VolumeSlider*)[super initWithWebView:theWebView];
    return self;
}

- (void)dealloc
{	
	[mpVolumeViewParentView release];
	[myVolumeView release];
    [super dealloc];
}


#pragma mark -
#pragma mark VolumeSlider

- (IBAction) sliderValueChanged:(UISlider *)sender {
   // NSLog(@"%.1f", [sender value]);
    NSString * jsCallBack = [NSString stringWithFormat:@"window.plugins.volumeSlider._onSliderChanged(%.1f, %d);", [sender value], [sender tag]];
    [self.webView stringByEvaluatingJavaScriptFromString:jsCallBack];
}

- (IBAction) sliderDragged:(UISlider *)sender {
    // NSLog(@"%.1f", [sender value]);
    NSString * jsCallBack = [NSString stringWithFormat:@"window.plugins.volumeSlider._onSliderDragged(%.1f, %d);", [sender value], [sender tag]];
    [self.webView stringByEvaluatingJavaScriptFromString:jsCallBack];
}

- (void) resize:(NSMutableArray*)arguments withDict:(NSMutableDictionary*)options
{
	self.callbackId = arguments.pop;
	NSUInteger argc = [arguments count];
	
	if (argc < 3) { // at a minimum we need x origin, y origin and width...
		return;
	}
	
	CGFloat originx,originy,width;
    int sliderId;
	CGFloat height = 30;
	
	originx = [[arguments objectAtIndex:0] floatValue];
	originy = [[arguments objectAtIndex:1] floatValue];
	width = [[arguments objectAtIndex:2] floatValue];
	if (argc < 4) {
		height = [[arguments objectAtIndex:3] floatValue];
	}
    
    sliderId = [[arguments objectAtIndex:4] intValue];
    
	CGRect viewRect = CGRectMake(
								 originx,
								 originy,
								 width,
								 height
								 );
    
            UIView *sliderView = (UIView *)[[self.webView superview] viewWithTag:(sliderId)];
    
    if ( sliderView ) {
        

        sliderView.frame = viewRect;
        [sliderView setAutoresizesSubviews:YES ];

        
        UISlider *slider = (UISlider *)[[self.webView superview] viewWithTag:(sliderId+200)];
        [slider setAutoresizingMask:( UIViewAutoresizingFlexibleWidth |
                                     UIViewAutoresizingFlexibleHeight )];

        
    }
    

}

- (void) createVolumeSlider:(NSMutableArray*)arguments withDict:(NSMutableDictionary*)options
{	
	self.callbackId = arguments.pop;
	NSUInteger argc = [arguments count];
	
	if (argc < 3) { // at a minimum we need x origin, y origin and width...
		return;	
	}
	
	CGFloat originx,originy,width;
    int sliderId;
	CGFloat height = 30;
	
	originx = [[arguments objectAtIndex:0] floatValue];
	originy = [[arguments objectAtIndex:1] floatValue];
	width = [[arguments objectAtIndex:2] floatValue];
	if (argc < 4) {
		height = [[arguments objectAtIndex:3] floatValue];
	}
    
    sliderId = [[arguments objectAtIndex:4] intValue];
    
	CGRect viewRect = CGRectMake(
								 originx, 
								 originy, 
								 width, 
								 height
								 );
	self.mpVolumeViewParentView = [[[UIView alloc] initWithFrame:viewRect] autorelease];

	[self.webView.superview addSubview:mpVolumeViewParentView];
	
	mpVolumeViewParentView.backgroundColor = [UIColor clearColor];
	self.myVolumeView = [[UISlider alloc] initWithFrame: mpVolumeViewParentView.bounds];

    
    mpVolumeViewParentView.tag = sliderId;
    [myVolumeView addTarget:self
               action:@selector(sliderValueChanged:)
     forControlEvents:UIControlEventTouchUpInside];
    
    [myVolumeView addTarget:self
                     action:@selector(sliderDragged:)
           forControlEvents:UIControlEventValueChanged];
    myVolumeView.tag = (sliderId+200);
    myVolumeView.minimumValue = 0;
    myVolumeView.maximumValue = 100;
	[mpVolumeViewParentView addSubview: myVolumeView];

}

- (void)showVolumeSlider:(NSMutableArray*)arguments withDict:(NSMutableDictionary*)options
{
    CGFloat sliderId = [[arguments objectAtIndex:1] floatValue];
    [self.mpVolumeViewParentView viewWithTag:sliderId].hidden = NO;
    
}

- (void)hideVolumeSlider:(NSMutableArray*)arguments withDict:(NSMutableDictionary*)options
{
    CGFloat sliderId = [[arguments objectAtIndex:1] floatValue];
    [self.mpVolumeViewParentView viewWithTag:sliderId].hidden = YES;
}

- (void)setVolumeSlider:(NSMutableArray*)arguments withDict:(NSMutableDictionary*)options
{
    if ([arguments objectAtIndex:1] > 0) {
        float val = [[arguments objectAtIndex:1] floatValue];
        int sliderId = [[arguments objectAtIndex:2] intValue];
        UISlider *slider = (UISlider *)[[self.webView superview] viewWithTag:(sliderId+200)];
        [slider setValue:val animated:YES];
        
    }

}


@end

File: www/scripts/phonegap.plugins/VolumeSlider.js
// Based off VolumeSlider.js by Tommy-Carlos Williams
//	
//	Slider Cordova Plugin
//      MIT Licensed
//
(function() {
	var cordovaRef = window.PhoneGap || window.Cordova || window.cordova; // old to new fallbacks
	var changeFuction = [];
	var dragFuction = [];
	var VolumeSlider = function() {};

	/**
	 * Create a volume slider.
	 */
	VolumeSlider.prototype.createVolumeSlider = function(originx, originy, width, height, tag) {
		cordovaRef.exec(null, null, "VolumeSlider", "createVolumeSlider", [originx, originy, width, height, tag]);
	};

	/**
	 * Show the volume slider
	 */
	VolumeSlider.prototype.showVolumeSlider = function(tag) {
		cordovaRef.exec(null, null, "VolumeSlider", "showVolumeSlider", [tag]);
	};
	/**
	 * Hide the volume slider
	 */
	VolumeSlider.prototype.hideVolumeSlider = function(tag) {
		cordovaRef.exec(null, null, "VolumeSlider", "hideVolumeSlider", [tag]);
	};
	/**
	 * Set the volume slider
	 */
	VolumeSlider.prototype.setVolumeSlider = function(val, tag) {
		if (isNaN(val) === false) {
			cordovaRef.exec(null, null, "VolumeSlider", "setVolumeSlider", [val, tag]);
		}
	};
	/**
	 * resize
	 */
	VolumeSlider.prototype.resize = function(originx, originy, width, height, tag) {
		cordovaRef.exec(null, null, "VolumeSlider", "resize", [originx, originy, width, height, tag]);
	};

	/**
	 * changed callback
	 */
	VolumeSlider.prototype._onSliderChanged = function(percentage, tag) {
		tag = tag - 200;
		if (changeFuction[tag])
			changeFuction[tag](percentage, tag);
	};

	VolumeSlider.prototype.onSliderChanged = function(callBkFunc, tag) {
		changeFuction[tag] = callBkFunc;
	};

	/**
	 * dragged callback
	 */
	VolumeSlider.prototype._onSliderDragged = function(percentage, tag) {
		tag = tag - 200;
		if (dragFuction[tag])
			dragFuction[tag](percentage, tag);
	};

	VolumeSlider.prototype.onSliderDragged = function(callBkFunc, tag) {
		dragFuction[tag] = callBkFunc;
	};

	cordovaRef.addConstructor(function() {
		if (!window.plugins) {
			window.plugins = {};
		}
		window.plugins.volumeSlider = new VolumeSlider();
	});
})();